modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1836,11 +1837,13 @@ private CompletableFuture<Void> handleChangePendingAssignmentEvent(
                     )
                     .thenComposeAsync(unused -> inBusyLock(busyLock, () -> {
                         if (!isRecovery) {
+                            // TODO https://issues.apache.org/jira/browse/IGNITE-21738 Fix potential race.
+                            HybridTimestamp lwm = lowWatermark.getLowWatermark();
                             // We create index storages (and also register the necessary structures) for the rebalancing one partition
                             // before start the raft node, so that the updates that come when applying the replication log can safely
                             // update the indexes. On recovery node, we do not need to call this code, since during restoration we start
                             // all partitions and already register indexes there.
-                            registerIndexesToTable(tbl, catalogService, singlePartitionIdSet, tbl.schemaView());
+                            registerIndexesToTable(tbl, catalogService, singlePartitionIdSet, tbl.schemaView(), lwm);
```

Comment:
What if you temporarily make reading the lwm under lock? 
Make some kind of method `LowWatermark#getLowWatermarkSafe(Consumer<HybridTimestamp> consumer)` in which it will be possible to safely read the sign?

Reply:
Done.